File: app/src/main/java/io/legado/app/api/ReaderProvider.java
/*
 * Copyright (C) 2020 w568w
 */

package io.legado.app.api;

import android.content.ContentProvider;
import android.content.ContentResolver;
import android.content.ContentValues;
import android.content.UriMatcher;
import android.database.CharArrayBuffer;
import android.database.ContentObserver;
import android.database.Cursor;
import android.database.DataSetObserver;
import android.net.Uri;
import android.os.Bundle;

import com.google.gson.Gson;

import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.Map;

import io.legado.app.web.controller.BookshelfController;
import io.legado.app.web.controller.SourceController;
import io.legado.app.web.utils.ReturnData;

/**
 * Export book data to other app.
 */
public class ReaderProvider extends ContentProvider {
    private enum RequestCode {
        saveSource, saveSources, saveBook, deleteSources, getSource, getSources, getBookshelf, getChapterList, getBookContent
    }

    public static final String POST_BODY_KEY = "json";
    public static final String AUTHORITY = "io.legado.app.api.ReaderProvider";
    private static UriMatcher sMatcher = new UriMatcher(UriMatcher.NO_MATCH);
    private SourceController mSourceController;
    private BookshelfController mBookshelfController;

    static {
        sMatcher.addURI(AUTHORITY, "source/insert", RequestCode.saveSource.ordinal());
        sMatcher.addURI(AUTHORITY, "sources/insert", RequestCode.saveSources.ordinal());
        sMatcher.addURI(AUTHORITY, "book/insert", RequestCode.saveBook.ordinal());
        sMatcher.addURI(AUTHORITY, "sources/delete", RequestCode.deleteSources.ordinal());
        sMatcher.addURI(AUTHORITY, "source/query", RequestCode.getSource.ordinal());
        sMatcher.addURI(AUTHORITY, "sources/query", RequestCode.getSources.ordinal());
        sMatcher.addURI(AUTHORITY, "books/query", RequestCode.getBookshelf.ordinal());
        sMatcher.addURI(AUTHORITY, "book/chapter/query", RequestCode.getChapterList.ordinal());
        sMatcher.addURI(AUTHORITY, "book/content/query", RequestCode.getBookContent.ordinal());
    }

    public ReaderProvider() {
    }

    @Override
    public int delete(@NotNull Uri uri, String selection, String[] selectionArgs) {
        if (sMatcher.match(uri) < 0)
            return -1;
        switch (RequestCode.values()[sMatcher.match(uri)]) {
            case deleteSources:
                mSourceController.deleteSources(selection);
                break;
            default:
                throw new IllegalStateException("Unexpected value: " + RequestCode.values()[sMatcher.match(uri)].name());
        }
        return 0;
    }

    @Override
    public String getType(@NotNull Uri uri) {
        throw new UnsupportedOperationException("Not yet implemented");
    }

    @Override
    public Uri insert(@NotNull Uri uri, ContentValues values) {
        if (sMatcher.match(uri) < 0)
            return null;
        switch (RequestCode.values()[sMatcher.match(uri)]) {
            case saveSource:
                mSourceController.saveSource(values.getAsString(POST_BODY_KEY));
                break;
            case saveBook:
                mBookshelfController.saveBook(values.getAsString(POST_BODY_KEY));
            case saveSources:
                mSourceController.saveSources(values.getAsString(POST_BODY_KEY));
            default:
                throw new IllegalStateException("Unexpected value: " + RequestCode.values()[sMatcher.match(uri)].name());
        }
        return null;
    }

    @Override
    public boolean onCreate() {
        mSourceController = new SourceController();
        mBookshelfController = new BookshelfController();
        return false;
    }

    @Override
    public Cursor query(Uri uri, String[] projection, String selection,
                        String[] selectionArgs, String sortOrder) {

        Map<String, ArrayList<String>> map = new HashMap<>();

        ArrayList<String> list = new ArrayList<>();
        list.add(uri.getQueryParameter("url"));
        map.put("url", list);

        list = new ArrayList<>();
        list.add(uri.getQueryParameter("index"));
        map.put("index", list);

        if (sMatcher.match(uri) < 0)
            return null;
        switch (RequestCode.values()[sMatcher.match(uri)]) {
            case getSource:
                return new SimpleCursor(mSourceController.getSource(map));
            case getSources:
                return new SimpleCursor(mSourceController.getSources());
            case getBookshelf:
                return new SimpleCursor(mBookshelfController.getBookshelf());
            case getBookContent:
                return new SimpleCursor(mBookshelfController.getBookContent(map));
            case getChapterList:
                return new SimpleCursor(mBookshelfController.getChapterList(map));
            default:
                throw new IllegalStateException("Unexpected value: " + RequestCode.values()[sMatcher.match(uri)].name());
        }
    }

    @Override
    public int update(@NotNull Uri uri, ContentValues values, String selection,
                      String[] selectionArgs) {
        throw new UnsupportedOperationException("Not yet implemented");
    }

    /**
     * Simple inner class to deliver json callback data.
     *
     * Only getString() makes sense.
     */
    private static class SimpleCursor implements Cursor {
        private String mData;

        public SimpleCursor(ReturnData data) {
            this.mData = new Gson().toJson(data);
        }

        @Override
        public int getCount() {
            return 1;
        }

        @Override
        public int getPosition() {
            return 0;
        }

        @Override
        public boolean move(int i) {
            return true;
        }

        @Override
        public boolean moveToPosition(int i) {
            return true;
        }

        @Override
        public boolean moveToFirst() {
            return true;
        }

        @Override
        public boolean moveToLast() {
            return true;
        }

        @Override
        public boolean moveToNext() {
            return true;
        }

        @Override
        public boolean moveToPrevious() {
            return true;
        }

        @Override
        public boolean isFirst() {
            return true;
        }

        @Override
        public boolean isLast() {
            return true;
        }

        @Override
        public boolean isBeforeFirst() {
            return false;
        }

        @Override
        public boolean isAfterLast() {
            return false;
        }

        @Override
        public int getColumnIndex(String s) {
            return 0;
        }

        @Override
        public int getColumnIndexOrThrow(String s) throws IllegalArgumentException {
            throw new UnsupportedOperationException("Not yet implemented");
        }

        @Override
        public String getColumnName(int i) {
            return null;
        }

        @Override
        public String[] getColumnNames() {
            return new String[0];
        }

        @Override
        public int getColumnCount() {
            return 0;
        }

        @Override
        public byte[] getBlob(int i) {
            return new byte[0];
        }

        @Override
        public String getString(int i) {
            return mData;
        }

        @Override
        public void copyStringToBuffer(int i, CharArrayBuffer charArrayBuffer) {

        }

        @Override
        public short getShort(int i) {
            return 0;
        }

        @Override
        public int getInt(int i) {
            return 0;
        }

        @Override
        public long getLong(int i) {
            return 0;
        }

        @Override
        public float getFloat(int i) {
            return 0;
        }

        @Override
        public double getDouble(int i) {
            return 0;
        }

        @Override
        public int getType(int i) {
            return 0;
        }

        @Override
        public boolean isNull(int i) {
            return false;
        }

        @Override
        public void deactivate() {

        }

        @Override
        public boolean requery() {
            return false;
        }

        @Override
        public void close() {

        }

        @Override
        public boolean isClosed() {
            return false;
        }

        @Override
        public void registerContentObserver(ContentObserver contentObserver) {

        }

        @Override
        public void unregisterContentObserver(ContentObserver contentObserver) {

        }

        @Override
        public void registerDataSetObserver(DataSetObserver dataSetObserver) {

        }

        @Override
        public void unregisterDataSetObserver(DataSetObserver dataSetObserver) {

        }

        @Override
        public void setNotificationUri(ContentResolver contentResolver, Uri uri) {

        }

        @Override
        public Uri getNotificationUri() {
            return null;
        }

        @Override
        public boolean getWantsAllOnMoveCalls() {
            return false;
        }

        @Override
        public void setExtras(Bundle bundle) {

        }

        @Override
        public Bundle getExtras() {
            return null;
        }

        @Override
        public Bundle respond(Bundle bundle) {
            return null;
        }
    }
}

File: app/src/main/java/io/legado/app/api/ReaderProvider.kt
/*
 * Copyright (C) 2020 w568w
 */
package io.legado.app.api

import android.content.ContentProvider
import android.content.ContentValues
import android.content.UriMatcher
import android.database.Cursor
import android.database.MatrixCursor
import android.net.Uri
import com.google.gson.Gson
import io.legado.app.api.controller.BookController
import io.legado.app.api.controller.BookSourceController
import io.legado.app.api.controller.RssSourceController
import kotlinx.coroutines.runBlocking

/**
 * Export book data to other app.
 */
class ReaderProvider : ContentProvider() {
    private enum class RequestCode {
        SaveBookSource, SaveBookSources, DeleteBookSources, GetBookSource, GetBookSources,
        SaveRssSource, SaveRssSources, DeleteRssSources, GetRssSource, GetRssSources,
        SaveBook, GetBookshelf, RefreshToc, GetChapterList, GetBookContent, GetBookCover,
        SaveBookProgress
    }

    private val postBodyKey = "json"
    private val sMatcher by lazy {
        UriMatcher(UriMatcher.NO_MATCH).apply {
            "${context?.applicationInfo?.packageName}.readerProvider".also { authority ->
                addURI(authority, "bookSource/insert", RequestCode.SaveBookSource.ordinal)
                addURI(authority, "bookSources/insert", RequestCode.SaveBookSources.ordinal)
                addURI(authority, "bookSources/delete", RequestCode.DeleteBookSources.ordinal)
                addURI(authority, "bookSource/query", RequestCode.GetBookSource.ordinal)
                addURI(authority, "bookSources/query", RequestCode.GetBookSources.ordinal)
                addURI(authority, "rssSource/insert", RequestCode.SaveBookSource.ordinal)
                addURI(authority, "rssSources/insert", RequestCode.SaveBookSources.ordinal)
                addURI(authority, "rssSources/delete", RequestCode.DeleteBookSources.ordinal)
                addURI(authority, "rssSource/query", RequestCode.GetBookSource.ordinal)
                addURI(authority, "rssSources/query", RequestCode.GetBookSources.ordinal)
                addURI(authority, "book/insert", RequestCode.SaveBook.ordinal)
                addURI(authority, "books/query", RequestCode.GetBookshelf.ordinal)
                addURI(authority, "book/refreshToc/query", RequestCode.RefreshToc.ordinal)
                addURI(authority, "book/chapter/query", RequestCode.GetChapterList.ordinal)
                addURI(authority, "book/content/query", RequestCode.GetBookContent.ordinal)
                addURI(authority, "book/cover/query", RequestCode.GetBookCover.ordinal)
            }
        }
    }

    override fun onCreate(): Boolean {
        context?.let { context ->
            ShortCuts.buildShortCuts(context)
        }
        return false
    }

    override fun delete(
        uri: Uri,
        selection: String?,
        selectionArgs: Array<String>?
    ): Int {
        if (sMatcher.match(uri) < 0) return -1
        when (RequestCode.entries[sMatcher.match(uri)]) {
            RequestCode.DeleteBookSources -> BookSourceController.deleteSources(selection)
            RequestCode.DeleteRssSources -> BookSourceController.deleteSources(selection)
            else -> throw IllegalStateException(
                "Unexpected value: " + RequestCode.entries[sMatcher.match(uri)].name
            )
        }
        return 0
    }

    override fun getType(uri: Uri) = throw UnsupportedOperationException("Not yet implemented")

    override fun insert(uri: Uri, values: ContentValues?): Uri? {
        if (sMatcher.match(uri) < 0) return null
        runBlocking {
            when (RequestCode.entries[sMatcher.match(uri)]) {
                RequestCode.SaveBookSource -> values?.let {
                    BookSourceController.saveSource(values.getAsString(postBodyKey))
                }

                RequestCode.SaveBookSources -> values?.let {
                    BookSourceController.saveSources(values.getAsString(postBodyKey))
                }

                RequestCode.SaveRssSource -> values?.let {
                    RssSourceController.saveSource(values.getAsString(postBodyKey))
                }

                RequestCode.SaveRssSources -> values?.let {
                    RssSourceController.saveSources(values.getAsString(postBodyKey))
                }

                RequestCode.SaveBook -> values?.let {
                    BookController.saveBook(values.getAsString(postBodyKey))
                }

                RequestCode.SaveBookProgress -> values?.let {
                    BookController.saveBookProgress(values.getAsString(postBodyKey))
                }

                else -> throw IllegalStateException(
                    "Unexpected value: " + RequestCode.entries[sMatcher.match(uri)].name
                )
            }
        }
        return null
    }

    override fun query(
        uri: Uri, projection: Array<String>?, selection: String?,
        selectionArgs: Array<String>?, sortOrder: String?
    ): Cursor? {
        val map: MutableMap<String, ArrayList<String>> = HashMap()
        uri.getQueryParameter("url")?.let {
            map["url"] = arrayListOf(it)
        }
        uri.getQueryParameter("index")?.let {
            map["index"] = arrayListOf(it)
        }
        uri.getQueryParameter("path")?.let {
            map["path"] = arrayListOf(it)
        }
        return if (sMatcher.match(uri) < 0) null else when (RequestCode.entries[sMatcher.match(uri)]) {
            RequestCode.GetBookSource -> SimpleCursor(BookSourceController.getSource(map))
            RequestCode.GetBookSources -> SimpleCursor(BookSourceController.sources)
            RequestCode.GetRssSource -> SimpleCursor(RssSourceController.getSource(map))
            RequestCode.GetRssSources -> SimpleCursor(RssSourceController.sources)
            RequestCode.GetBookshelf -> SimpleCursor(BookController.bookshelf)
            RequestCode.GetBookContent -> SimpleCursor(BookController.getBookContent(map))
            RequestCode.RefreshToc -> SimpleCursor(BookController.refreshToc(map))
            RequestCode.GetChapterList -> SimpleCursor(BookController.getChapterList(map))
            RequestCode.GetBookCover -> SimpleCursor(BookController.getCover(map))
            else -> throw IllegalStateException(
                "Unexpected value: " + RequestCode.entries[sMatcher.match(uri)].name
            )
        }
    }

    override fun update(
        uri: Uri, values: ContentValues?, selection: String?,
        selectionArgs: Array<String>?
    ) = throw UnsupportedOperationException("Not yet implemented")


    /**
     * Simple inner class to deliver json callback data.
     *
     * Only getString() makes sense.
     */
    private class SimpleCursor(data: ReturnData?) : MatrixCursor(arrayOf("result"), 1) {

        private val mData: String = Gson().toJson(data)

        init {
            addRow(arrayOf(mData))
        }

    }
}